### PR TITLE
Refactor `S3Bucket` block

### PR DESCRIPTION
**Pull request by ahuang11**
Update S3Bucket to make it compliant with ObjectStorageBlock, and refactor a lot along the way, like making AwsClientParameters modifyable on the UI and merging MinIOCredentials into aws_credentials field.

Closes a bunch
https://github.com/PrefectHQ/prefect-aws/issues/149
https://github.com/PrefectHQ/prefect-aws/issues/150
https://github.com/PrefectHQ/prefect-aws/issues/146
https://github.com/PrefectHQ/prefect-aws/issues/162
https://github.com/PrefectHQ/prefect-aws/issues/124

### Example

See readme

### Screenshots

image
image

### Checklist

- [x] References any related issue by including "Closes #<Issue Number>" or "Closes <Issue URL>".
  - If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect-aws/issues/new/choose) first.
- [x] Includes tests or only affects documentation.
- [x] Passes `pre-commit` checks.
  - Run `pre-commit install && pre-commit run --all` locally for formatting and linting.
- [x] Includes screenshots of documentation updates.
  - Run `mkdocs serve` view documentation locally.
- [x] Summarizes PR's changes in [CHANGELOG.md](https://github.com/PrefectHQ/prefect-aws/blob/main/CHANGELOG.md)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,10 +7,21 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
 
+### Added
+
+- `list_objects`, `download_object_to_path`, `download_object_to_file_object`, `download_folder_to_path`, `upload_from_path`, `upload_from_file_object`, `upload_from_folder` methods in `S3Bucket` - [#85](https://github.com/PrefectHQ/prefect-gcp/pull/175)
+
 ### Changed
 
+- `S3Bucket` additionally inherits from `ObjectStorageBlock` - [#175](https://github.com/PrefectHQ/prefect-gcp/pull/175)
+- Exposed all existing blocks to the top level init - [#175](https://github.com/PrefectHQ/prefect-gcp/pull/175)
+
```
Comment (ahuang11) [inline CHANGELOG.md:18]: Probably should mention aws_client_parameters as a field to AwsCredentials and MinioCredentials

File: tests/test_s3_bucket.py [hunk 2]
```diff
@@ -1,331 +0,0 @@
-from pathlib import Path
```
Comment (ahuang11) [inline tests/test_s3_bucket.py:1]: I migrated the tests to test_s3.py

File: requirements-dev.txt [hunk 3]
```diff
@@ -17,3 +17,4 @@ coverage
 types-boto3 >= 1.0.2
 boto3-stubs >= 1.24.39
 pillow
+mypy_boto3_s3
```
Comment (ahuang11) [inline requirements-dev.txt:20]: Better typing

File: prefect_aws/s3.py [hunk 4]
```diff
@@ -261,23 +264,39 @@ class S3Bucket(WritableFileSystem, WritableDeploymentStorage):
...
-        description="A block containing your credentials (choose this or "
-        "MinIO Credentials)",
+        description=(
+            "[DEPRECATED; use the aws_credentials field instead] "
+            "A block containing your credentials (choose this or "
+            "AWS Credentials)"
+        ),
     )
-    aws_credentials: Optional[AwsCredentials] = Field(
+    aws_credentials: Optional[Union[AwsCredentials, MinIOCredentials]] = Field(
```
Comment (ahuang11) [inline prefect_aws/s3.py:273]: The UI doesn't seem to work with this...

File: prefect_aws/s3.py [hunk 5]
```diff
@@ -261,23 +264,39 @@ class S3Bucket(WritableFileSystem, WritableDeploymentStorage):
     bucket_name: str = Field(default=..., description="Name of your bucket")
     minio_credentials: Optional[MinIOCredentials] = Field(
         default=None,
-        description="A block containing your credentials (choose this or "
-        "MinIO Credentials)",
+        description=(
+            "[DEPRECATED; use the aws_credentials field instead] "
```
Comment (ahuang11) [inline prefect_aws/s3.py:268]: Probably needs to be in docstring too or remove the docstring

File: prefect_aws/s3.py [hunk 6]
```diff
@@ -576,3 +631,437 @@ def _write_sync(self, key: str, data: bytes) -> None:
...
+
+        # do not call self._join_bucket_folder for filter
+        # because it's built-in to that method already!
+        # however, we still need to do it because we're using relative_to
+        bucket_folder = self._join_bucket_folder(from_folder)
+
+        async_coros = []
+        for object in objects:
+            bucket_path = Path(object["Key"]).relative_to(bucket_folder)
+            if bucket_path.is_dir():
```
Comment (zzstoatzz) [inline prefect_aws/s3.py:880]: does this mean that any directories (which may themselves contain files) within `from_folder` will not be downloaded?

Comment (ahuang11) [reply prefect_aws/s3.py:880]: Sub-directories will be downloaded; it's only if it's literally a directory.

I can add a comment.

**Review by desertaxle — CHANGES_REQUESTED** [review]
It looks like we're adding a fair amount of code to handle deprecated fields which can be removed once the deprecated fields are removed. Can you add some `#TODO:` comments to help us keep track of what code should be removed when deprecated fields are removed?

**Comment by ahuang11** [comment]
image

**Review by desertaxle — APPROVED** [review]
LGTM!

**Pull request merged**